### PR TITLE
Enforce union case declarations AttributeTargets

### LINKED ISSUE #3112: Compiler ignores AttributeUsage for Attributes attached to union cases
While browsing stackoverflow for something unrelated, I found an old mention of some compiler error.

(ref https://stackoverflow.com/questions/24884814/serializing-f-discriminated-unions-with-protobuf)

```F#

open System

[<AttributeUsage(AttributeTargets.Method)>]
type MethodLevelAttribute() =
    inherit Attribute()

[<AttributeUsage(AttributeTargets.Property ||| AttributeTargets.Field)>]
type PropertyOrFieldLevelAttribute() =
    inherit Attribute()

// correctly fails
//[<MethodLevelAttribute>]
//type SomeType() = class end

type SomeUnion =
| [<PropertyOrFieldLevelAttribute>] Case1 of int
| [<PropertyOrFieldLevelAttribute>] Case2 of string

module Program =

    [<EntryPoint>]
    let main argv = 
        printfn "%A" argv
        0 // return an integer exit code

```

## Expected:

Compiler Error

## Actual:

The compiler stuck the attributes to the static construction methods:

image

## Version: 

```<package id="FSharp.Compiler.Tools" version="4.1.17" targetFramework="net452" />```
(also repros with current nightly)

## Note:

I only checked this one case, there may be other cases where it is also not correctly checked.

### PR DESCRIPTION
**Pull request by edgarfgp**
## Description

Fixes #3112

UnionCase compiles down to a static `Method` so  should only enforce  `AttributeTargets.Method` [SharpLab](https://sharplab.io/#v2:DYLgZgzgNAJiDUAfA9gBwKYDsAEBlAnhAC7oC2AsAFBUDaAPAIJFEBOAlgEYCuJAqhAEMA5ugAUTVpx7oAKgJYiiEAHQAFFmnQsi+bIn3YJ7biTkL0S5QDE26YDACUAPgC6VHRmzrN2/AHkWGzsYABl0ADc7IykSUQdsAF4qbBTsNkwACy02IkNmY2k45NTaRnyY9H5hMWiTWXlFFQBZCwzkR1d3fE8WojbQiKjyurjE4pT0rPZc2sKHKi7PXGRSSsw2ZBwkykRsem8MXwCg+zDI4FmSV2wAYQEIdABGbGQwNMwiKl36Xv6zu2udweACYXm9iOxMEIqEA===)

### Old Rule
- AttributeTargets.Method ||| AttributeTargets.Property

### New Rule
- AttributeTargets.Method

## Checklist

- [x] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by edgarfgp** [comment]
This is ready

File: src/Compiler/Checking/CheckExpressions.fs [hunk 1]
```diff
@@ -10874,17 +10874,23 @@ and TcNormalizedBinding declKind (cenv: cenv) env tpenv overallTy safeThisValOpt
...
-
-            TcAttributesWithPossibleTargets false cenv env attrTgt attrs |> ignore
+            TcAttributeTargetsOnLetBindings cenv env attrs overallPatTy overallExprTy declaredTypars.IsEmpty
 
         CheckedBindingInfo(inlineFlag, valAttribs, xmlDoc, tcPatPhase2, explicitTyparInfo, nameToPrelimValSchemeMap, rhsExprChecked, argAndRetAttribs, overallPatTy, mBinding, debugPoint, isCompGen, literalValue, isFixed), tpenv
 
+// Note:
+// - Let bound values can only have attributes that uses AttributeTargets.Field ||| AttributeTargets.Property ||| AttributeTargets.ReturnValue
+// - Let function bindings can only have attributes that uses AttributeTargets.Method ||| AttributeTargets.ReturnValue
+and TcAttributeTargetsOnLetBindings (cenv: cenv) env attrs overallPatTy overallExprTy hasDeclTypars =
```
Comment (psfinaki) [inline src/Compiler/Checking/CheckExpressions.fs:10884]: Shouldn't this be called the opposite here? `hasDeclTypars = declaredTypars.IsEmpty` here :)
Or maybe the logic can be inverted/negated, to keep the names clear.

Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs]: Sure. Thanks for the review

**Review by psfinaki — APPROVED** [review]
Thanks for the fix, tests and refactoring. That's a good one.

I wonder if we should add some notes to some docs about it. This is a technically correct behavior but might seem unintuitive to F# newcomers.

**Comment by edgarfgp** [comment]
> Thanks for the fix, tests and refactoring. That's a good one.
> 
> I wonder if we should add some notes to some docs about it. This is a technically correct behavior but might seem unintuitive to F# newcomers.

Im planning to submit a RFC with the `AttributeTargets` rules once Im done with https://github.com/dotnet/fsharp/issues/13526 to make it clear

**Comment by psfinaki** [comment]
Awesome :)

**Comment by edgarfgp** [comment]
This should merged before https://github.com/dotnet/fsharp/pull/16790. Thanks

**Comment by psfinaki** [comment]
@dotnet/fsharp-team-msft

File: src/Compiler/Checking/CheckExpressions.fs [hunk 2]
```diff
@@ -10874,17 +10874,23 @@ and TcNormalizedBinding declKind (cenv: cenv) env tpenv overallTy safeThisValOpt
...
+// Note:
+// - Let bound values can only have attributes that uses AttributeTargets.Field ||| AttributeTargets.Property ||| AttributeTargets.ReturnValue
+// - Let function bindings can only have attributes that uses AttributeTargets.Method ||| AttributeTargets.ReturnValue
+and TcAttributeTargetsOnLetBindings (cenv: cenv) env attrs overallPatTy overallExprTy areTyparsDeclared =
+    let rhsIsFunction = isFunTy cenv.g overallPatTy
+    let lhsIsFunction = isFunTy cenv.g overallExprTy
+    let attrTgt =
+        match rhsIsFunction, lhsIsFunction with
+        | false, false when not areTyparsDeclared -> AttributeTargets.Field ||| AttributeTargets.Property ||| AttributeTargets.ReturnValue
+        | _, _ -> AttributeTargets.Method ||| AttributeTargets.ReturnValue
```
Comment (brianrourkeboll) [inline src/Compiler/Checking/CheckExpressions.fs:10890]: I see that all you did was move this code, but even so, the logic might be a little easier to understand like this:

```suggestion
    let attrTgt =
        if
            // It's a type function:
            //     let x<'a> = …
            areTyparsDeclared
            // It's a regular function-valued binding:
            //     let f x = …
            //     let f = fun x -> …
            || isFunTy cenv.g overallPatTy
            || isFunTy cenv.g overallExprTy
        then
            AttributeTargets.ReturnValue ||| AttributeTargets.Method
        else
            AttributeTargets.ReturnValue ||| AttributeTargets.Field ||| AttributeTargets.Property
```

Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs]: Yeah much clear. Thanks for the review

**Comment by edgarfgp** [comment]
> That said, I don't mind merging it as it is now, but it still feels weird to me that we allow `Method` and not `Constructor`.

@vzarytovskii As you know there not an actual spec yet for this, so if `Constructor` make sense from the F# perspective we can add it

**Comment by auduchinok** [comment]
> UnionCase compiles down to a static Method

This is not always the case. There's no single direct mapping between union cases and their compiled .NET representations: depending on the union and its union cases, they compile to different things (nested classes, properties, constructor methods, case test properties, nested `Tags` enum and so on).

For example, these union cases compile to properties, not methods:
```fsharp
type U =
    | A
    | B
```

I don't see any tests covering these cases.

Is there an RFC for this?

**Comment by edgarfgp** [comment]
> > UnionCase compiles down to a static Method
> 
> This is not always the case. There's no single direct mapping between union cases and their compiled .NET representations: depending on the union and its union cases, they compile to different things (nested classes, properties, constructor methods, case test properties, nested `Tags` enum and so on).
> 
> For example, these union cases compile to properties, not methods:
> 
> ```fsharp
> type U =
>     | A
>     | B
> ```
> 
> I don't see any tests covering these cases.
> 
> Is there an RFC for this?

@auduchinok  No Yet. Im going through the issues and figuring out which `AttributeTargets` should be enforced. Will create a RFC at some point as soon I have a clear picture of this.

> This is not always the case. There's no single direct mapping between union cases and their compiled .NET representations: depending on the union and its union cases, they compile to different things (nested classes, properties, constructor methods, case test properties, nested `Tags` enum and so on).

I will look into this as create a PR with tests

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 3]
```diff
@@ -511,7 +511,14 @@ module TcRecdUnionAndEnumDeclarations =
                 
     let TcUnionCaseDecl (cenv: cenv) env parent thisTy thisTyInst tpenv hasRQAAttribute (SynUnionCase(Attributes synAttrs, SynIdent(id, _), args, xmldoc, vis, m, _)) =
         let g = cenv.g
-        let attrs = TcAttributes cenv env AttributeTargets.UnionCaseDecl synAttrs // the attributes of a union case decl get attached to the generated "static factory" method
+        let attrs =
+            // The attributes of a union case decl get attached to the generated "static factory" method
+            // Enforce that the union-cases can only be targeted by attributes with AttributeTargets.Method
+            if g.langVersion.SupportsFeature(LanguageFeature.EnforceAttributeTargetsUnionCaseDeclarations) then
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:517]: Following the logic from IlxGen, this should branch on arguments for the case.
Case with no data is a property.
Case with fields is a method.

There are other concerns affecting DU IL generation (structness, null as true value, number of cases), but so far I have not found any other affecting case-level attribute placement beyond what I wrote above (and what Eugene pointed out).

Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:517]: Thanks for the review. See https://github.com/dotnet/fsharp/pull/16807

Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:517]: If you find any other case. Let me know, I really want to improve this area of the compiler

**Pull request merged**